### PR TITLE
Adding images to the map with same interface as Places (via Plugin trait).

### PR DESCRIPTION
**Pull request by Nikitamuraviev10**
There are some artifacts present when rotating the image, but it's better than nothing.
"Adding tile caching to disk" was deleted during "Merge with main". Don't worry about that.

### CODE REVIEW

File: walkers/src/extras/images.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    /// Same as [egui::TextureHandle::size] (https://docs.rs/egui/latest/egui/struct.TextureHandle.html#method.size)
+    #[inline(always)]
+    pub fn size(&self) -> [usize; 2] {
+        self.texture.size()
+    }
+
+    /// Scale texture.
+    pub fn scale(&mut self, x_val: f32, y_val: f32) {
```
Comment (podusowski) [inline walkers/src/extras/images.rs:132]: Scaling is implemented somewhere in egui as well.

https://docs.rs/egui/latest/egui/widgets/struct.Image.html

**Review by podusowski — COMMENTED** [review]
I'll probably do more thorough review later, but I generally like the idea :)

File: walkers/src/extras/images.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            let h = h as f32;
+            let mut rect = map_rect.translate(screen_position);
+
+            rect.min.x -= w / 2.0;
+            rect.min.y -= h / 2.0;
+
+            rect.max.x = rect.min.x + w;
+            rect.max.y = rect.min.y + h;
+
+            let skip = (rect.max.x < map_rect.min.x)
```
[the remark this answers is not in the document]
Comment (Nikitamuraviev10) [reply walkers/src/extras/images.rs]: Redone in new commit.

File: walkers/src/extras/images.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    pub fn rotate(&mut self, degree: f32) {
+        if degree == self.angle {
+            return;
+        }
+        self.angle = degree;
+        self.do_rotate();
+    }
+
+    // This code based on transform::rotate from [raster] (https://crates.io/crates/raster) crate
+    fn do_rotate(&mut self) {
```
[the remark this answers is not in the document]
Comment (Nikitamuraviev10) [reply walkers/src/extras/images.rs:147]: Yes. I have seen this method. But I couldn't get it to work. I tried creating egui::Image/mesh from texture and passing it to painter.add (in draw method), but it does'n work.

[the remark this answers is not in the document]
Comment (Nikitamuraviev10) [reply walkers/src/extras/images.rs]: You are right! Redone in new commit.

File: walkers/src/extras/images.rs (hunk 1 above)
Comment (Nikitamuraviev10) [reply walkers/src/extras/images.rs:130]: Redone in new commit.

File: demo/src/lib.rs [hunk 4]
```diff
@@ -9,15 +9,29 @@ pub struct MyApp {
     geoportal_tiles: Tiles,
     map_memory: MapMemory,
     satellite: bool,
+    texture: Texture,
+    rotate: f32,
+    x_scale: f32,
+    y_scale: f32,
```
Comment (podusowski) [inline demo/src/lib.rs:15]: Could you group them into some struct, to point out that they are related to the images drawing part of the demo?

File: demo/src/lib.rs [hunk 5]
```diff
@@ -95,6 +120,11 @@ mod places {
     pub fn dworcowa_bus_stop() -> Position {
         Position::new(17.03940, 51.10005)
     }
+
+    /// Shopping center
```
Comment (podusowski) [inline demo/src/lib.rs:124]:
```suggestion
    /// Shopping center, and main intercity bus station.
```

File: Cargo.toml [hunk 6]
```diff
@@ -6,3 +6,4 @@ log = "0.4"
 egui = "0.22"
 egui_extras = { version = "0.22", features = ["image"] }
 eframe = "0.22"
+emath = "0.22"
```
Comment (podusowski) [inline Cargo.toml:9]: Egui re-exports emath in `egui::emath`. One less version to maintain.

File: walkers/src/extras/images.rs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            }
+        }
+    }
+}
+
+trait ByPixel {
+    fn get_pixel(&self, x: usize, y: usize) -> Option<Color32>;
+    fn set_pixel(&mut self, x: usize, y: usize, color: Color32);
+    fn index_to_xy(&self, i: usize) -> (usize, usize);
+}
```
Comment (podusowski) [inline walkers/src/extras/images.rs:73]: I suppose we don't need it now.

File: walkers/src/extras/images.rs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    #[inline(always)]
+    pub fn scale(&mut self, x_val: f32, y_val: f32) {
+        self.x_scale = x_val;
+        self.y_scale = y_val;
+    }
+
+    /// Rotate texture.
+    /// Angle is clockwise in radians. A 𝞃/4 = 90° rotation means rotating the X axis to the Y axis.
+    #[inline(always)]
+    pub fn rotate(&mut self, angle: f32) {
```
Comment (podusowski) [inline walkers/src/extras/images.rs:138]:
```suggestion
    pub fn angle(&mut self, angle: f32) {
```

Rotate suggests that if you do it twice, the image will be rotate twice as much.

**Pull request merged**